pkg/download/blobwithmsitoken.go

```diff
@@ -0,0 +1,98 @@
+package download
+
+import (
+	"fmt"
+	"github.com/Azure/azure-extension-foundation/httputil"
+	"github.com/Azure/azure-extension-foundation/msi"
+	"github.com/pkg/errors"
+	"net/http"
+	url2 "net/url"
+	"strings"
+)
+
+const (
+	xMsVersionHeaderName = "x-ms-version"
+	xMsVersionValue      = "2018-03-28"
+	storageResourceName  = "https://storage.azure.com/"
+)
+
+var azureBlobDomains = map[string]interface{}{ // golang doesn't have builtin hash sets, so this is a workaround for that
+	"blob.core.windows.net":       nil,
+	"blob.core.chinacloudapi.cn":  nil,
+	"blob.core.usgovcloudapi.net": nil,
+	"blob.core.couldapi.de":       nil,
+}
+
+type blobWithMsiToken struct {
+	url         string
+	msiProvider MsiProvider
+}
+
+type MsiProvider func() (msi.Msi, error)
+
+func (self *blobWithMsiToken) GetRequest() (*http.Request, error) {
+	msi, err := self.msiProvider()
+	if err != nil {
+		return nil, err
+	}
+	if msi.AccessToken == "" {
+		return nil, errors.New("MSI token was empty")
+	}
+
+	request, err := http.NewRequest(http.MethodGet, self.url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if IsAzureStorageBlobUri(self.url) {
+		request.Header.Set("Authorization", fmt.Sprintf("Bearer %s", msi.AccessToken))
+		request.Header.Set(xMsVersionHeaderName, xMsVersionValue)
+	}
+	return request, nil
+}
+
+func NewBlobWithMsiDownload(url string, msiProvider MsiProvider) Downloader {
+	return &blobWithMsiToken{url, msiProvider}
+}
+
+func GetMsiProviderForStorageAccountsImplicitly(blobUri string) MsiProvider {
+	msiProvider := msi.NewMsiProvider(httputil.NewSecureHttpClient(httputil.DefaultRetryBehavior))
+	return func() (msi.Msi, error) { return msiProvider.GetMsiForResource(GetResourceNameFromBlobUri(blobUri)) }
+}
+
+func GetMsiProviderForStorageAccountsWithClientId(blobUri, clientId string) MsiProvider {
+	msiProvider := msi.NewMsiProvider(httputil.NewSecureHttpClient(httputil.DefaultRetryBehavior))
+	return func() (msi.Msi, error) {
+		return msiProvider.GetMsiUsingClientId(clientId, GetResourceNameFromBlobUri(blobUri))
+	}
+}
+
+func GetMsiProviderForStorageAccountsWithObjectId(blobUri, objectId string) MsiProvider {
+	msiProvider := msi.NewMsiProvider(httputil.NewSecureHttpClient(httputil.DefaultRetryBehavior))
+	return func() (msi.Msi, error) {
```

Comment:
should we check objectid is valid here ?

Reply:
Already checked as a part of JSON schema check

https://github.com/Azure/custom-script-extension-linux/blob/10710fb27d7fb5061d9193f3f6edbb8b82b1a5a8/main/schema.go#L84